lib/timers.js

```diff
@@ -554,67 +557,73 @@ function tryOnImmediate(immediate, queue) {
   }
 }
 
+function runCallback(timer) {
+  var argv = timer._argv;
+  var argc = argv ? argv.length : 0;
+  switch (argc) {
+    // fast-path callbacks with 0-3 arguments
+    case 0:
+      return timer._callback();
+    case 1:
+      return timer._callback(argv[0]);
+    case 2:
+      return timer._callback(argv[0], argv[1]);
+    case 3:
+      return timer._callback(argv[0], argv[1], argv[2]);
+    // more then 3 arguments run slower with .apply
+    default:
+      return timer._callback.apply(timer, argv);
+  }
+}
 
-function Immediate() { }
-
-Immediate.prototype.domain = undefined;
-Immediate.prototype._onImmediate = undefined;
-Immediate.prototype._idleNext = undefined;
-Immediate.prototype._idlePrev = undefined;
 
+function createImmediate(callback, args) {
+  return {
+    _idleNext: null,
+    _idlePrev: null,
+    _callback: callback,
+    _argv: args,
+    _onImmediate: callback,
+    domain: process.domain,
+  };
+}
 
 exports.setImmediate = function(callback, arg1, arg2, arg3) {
   if (typeof callback !== 'function') {
     throw new TypeError('"callback" argument must be a function');
   }
 
   var i, args;
-  var len = arguments.length;
-  var immediate = new Immediate();
-
-  L.init(immediate);
 
-  switch (len) {
+  switch (arguments.length) {
     // fast cases
     case 0:
     case 1:
-      immediate._onImmediate = callback;
       break;
     case 2:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1);
-      };
+      args = [arg1];
       break;
     case 3:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2);
-      };
+      args = [arg1, arg2];
       break;
     case 4:
-      immediate._onImmediate = function() {
-        callback.call(immediate, arg1, arg2, arg3);
-      };
+      args = [arg1, arg2, arg3];
       break;
     // slow case
     default:
+      var len = arguments.length;
```

Comment:
correct, `x = [1, 2, 3]` runs faster than `x = new Array(3)`, and setting the values is extra.
Some 50+% faster to create pre-populated arrays of fixed length 3.  (In fact, [1], [1,2,3] and
[1,2,3,4,5,6,7,8,9] all run equally fast, while explicitly setting values is linear in array length.)

Normally either would be blazing fast, 75 million vs 48 million / second, but when tuning
code that itself runs at 4 million operations / second, the difference amounts to 3% of the total.
(the baseline was 1 million / second, where the difference was under 1%)